### PR TITLE
Reconnect slasher streams on beacon node shutdown

### PR DESCRIPTION
**Pull request by shayzluf**
[Resolves] #5375

---

# Description
resolves slasher log spamming and try to reconnect beacon node and restart stream when stream is closed

### CODE REVIEW

File: slasher/beaconclient/receivers.go [hunk 1]
```diff
@@ -33,6 +39,13 @@ func (bs *Service) receiveBlocks(ctx context.Context) {
 			log.WithError(ctx.Err()).Error("Context canceled - shutting down blocks receiver")
 			return
 		}
+		if err != nil && strings.Contains(err.Error(), "Context canceled") {
```
Comment (0xKiwi) [inline slasher/beaconclient/receivers.go:42]: Can we maybe move the error text to a global var?

File: slasher/beaconclient/receivers.go [hunk 2]
```diff
@@ -120,3 +140,43 @@ func (bs *Service) collectReceivedAttestations(ctx context.Context) {
...
+	}
+
+}
+
+func (bs *Service) restartBlockStream(ctx context.Context) (ethpb.BeaconChain_StreamBlocksClient, error) {
+	ticker := time.NewTicker(ReconnectPeriod)
+	for {
+		select {
+		case <-ticker.C:
+			log.Info("Context closed, trying to restart block stream")
```
Comment (0xKiwi) [inline slasher/beaconclient/receivers.go:169]:
```suggestion
			log.Info("Context closed, attempting to restart block stream")
```

File: slasher/beaconclient/receivers.go [hunk 3]
```diff
@@ -120,3 +140,43 @@ func (bs *Service) collectReceivedAttestations(ctx context.Context) {
...
+			log.Info("Context closed, trying to restart attestation stream")
+			stream, err := bs.beaconClient.StreamIndexedAttestations(ctx, &ptypes.Empty{})
+			if err != nil {
+				continue
+			}
+			log.Info("Attestation stream restarted...")
+			return stream, nil
+		case <-ctx.Done():
+			log.Debug("Context closed, exiting reconnect routine")
+			return nil, errors.New("context closed, quit restart stream retries")
```
Comment (0xKiwi) [inline slasher/beaconclient/receivers.go:158]:
```suggestion
			return nil, errors.New("context closed, not longer attempting to restart stream")
```

File: slasher/beaconclient/receivers.go [hunk 4]
```diff
@@ -120,3 +140,43 @@ func (bs *Service) collectReceivedAttestations(ctx context.Context) {
 		}
 	}
 }
+
+func (bs *Service) restartIndexedAttestationStream(ctx context.Context) (ethpb.BeaconChain_StreamIndexedAttestationsClient, error) {
+	ticker := time.NewTicker(ReconnectPeriod)
+	for {
+		select {
+		case <-ticker.C:
+			log.Info("Context closed, trying to restart attestation stream")
```
Comment (0xKiwi) [inline slasher/beaconclient/receivers.go:149]:
```suggestion
			log.Info("Context closed, attempting to restart attestation stream")
```

File: slasher/beaconclient/receivers.go [hunk 5]
```diff
@@ -11,6 +13,10 @@ import (
 	"go.opencensus.io/trace"
 )
 
+// ReconnectPeriod represents the frequency in which we check the number of
+// open connections.
+var ReconnectPeriod = 5 * time.Second
```
Comment (0xKiwi) [inline slasher/beaconclient/receivers.go:18]: Does this need to be exported?

File: slasher/beaconclient/receivers.go [hunk 6]
```diff
@@ -120,3 +140,43 @@ func (bs *Service) collectReceivedAttestations(ctx context.Context) {
...
+			log.Info("Context closed, trying to restart block stream")
+			stream, err := bs.beaconClient.StreamBlocks(ctx, &ptypes.Empty{})
+			if err != nil {
+				continue
+			}
+			log.Info("Block stream restarted...")
+			return stream, nil
+		case <-ctx.Done():
+			log.Debug("Context closed, exiting reconnect routine")
+			return nil, errors.New("context closed, quit restart stream retries")
```
Comment (0xKiwi) [inline slasher/beaconclient/receivers.go:178]:
```suggestion
			return nil, errors.New("context closed, no longer attempting to restart stream")
```

File: slasher/beaconclient/receivers.go (hunk 1 above)
Comment (shayzluf) [reply slasher/beaconclient/receivers.go]: changed to context.Canceled.Error()

Comment (shayzluf) [reply slasher/beaconclient/receivers.go]: thanks

File: slasher/beaconclient/receivers.go (hunk 5 above)
Comment (shayzluf) [reply slasher/beaconclient/receivers.go]: changed to not exported

File: slasher/beaconclient/receivers.go [hunk 7]
```diff
@@ -11,6 +13,10 @@ import (
 	"go.opencensus.io/trace"
 )
 
+// represents the frequency in which we try to restart our
+// streams when beacon chain is down.
```
Comment (0xKiwi) [inline slasher/beaconclient/receivers.go:17]:
```suggestion
// reconnectPeriod is the frequency that we try to restart our
// streams when the beacon chain is node does not respond.
```

File: slasher/beaconclient/receivers.go [hunk 8]
```diff
@@ -33,6 +39,13 @@ func (bs *Service) receiveBlocks(ctx context.Context) {
 			log.WithError(ctx.Err()).Error("Context canceled - shutting down blocks receiver")
 			return
 		}
+		if err != nil && strings.Contains(strings.ToLower(err.Error()), strings.ToLower(context.Canceled.Error())) {
```
Comment (rauljordan) [inline slasher/beaconclient/receivers.go:42]: I think you can do
if ctx.Err() == context.Canceled {

File: slasher/beaconclient/receivers.go [hunk 9]
```diff
@@ -70,9 +83,16 @@ func (bs *Service) receiveAttestations(ctx context.Context) {
 			log.WithError(ctx.Err()).Error("Context canceled - shutting down attestations receiver")
 			return
 		}
+		if err != nil && strings.Contains(strings.ToLower(err.Error()), strings.ToLower(context.Canceled.Error())) {
```
Comment (rauljordan) [inline slasher/beaconclient/receivers.go:86]: +1

File: slasher/beaconclient/receivers.go (hunk 8 above)
Comment (shayzluf) [reply slasher/beaconclient/receivers.go:42]: i tried it and it doesn't work as the error is rpc error and context cancelled is and inner error with lower letters

File: slasher/beaconclient/receivers.go (hunk 9 above)
Comment (shayzluf) [reply slasher/beaconclient/receivers.go:86]: +1

Comment (shayzluf) [reply slasher/beaconclient/receivers.go]: fixed and tested by following https://stackoverflow.com/questions/52969205/how-to-assert-grpc-error-codes-client-side-in-go

File: slasher/beaconclient/receivers.go (hunk 8 above)
Comment (shayzluf) [reply slasher/beaconclient/receivers.go]: fixed.
thanks

**Pull request merged**